add a tracking table for course run purchases

rachellougee opened the pull request:
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/738)

#### What's this PR do?
 Add a new table to track when order is fulfilled or in review state. see https://github.com/mitodl/mitxonline/issues/738 for detail
 The insert logic is implemented in the following places
- When order is transition from pending or review to fulfilled
- When order is transition from pending to review

Note that 
- If order is manually updated to fulfilled or review from admin directly, It would not work.
- No delete is implemented

#### How should this be manually tested?
enroll a course from frontend, it should display this if order is paid successful
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/3138890/181248408-d2bd8153-8e65-4f69-8f82-ed859d582c77.png">
Go to admin/courses/paidcourserun/, you should see the fulfilled order shows up there
![image](https://user-images.githubusercontent.com/3138890/181248754-0f700698-85a6-4911-b722-8865f661da07.png)



#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

jkachel commented on ecommerce/api.py:
```diff
@@ -164,6 +167,21 @@ def fulfill_completed_order(
+    order = Order.objects.get(id=order.id)
+    if order.STATE.FULFILLED or order.STATE.REVIEW:
+        for run in order.purchased_runs:
+            if not PaidCourseRun.objects.filter(
+                order=order, course_run=run, user=order.purchaser
+            ).exists():
+                PaidCourseRun.objects.create(
+                    order=order, course_run=run, user=order.purchaser
+                )
```
Small change: there's a `update_or_create()` [function](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#update-or-create) that could be used here. (Or, for that matter, `get_or_create()` if you don't want it to update the timestamps.)

jkachel reviewed (changes requested):
One other change: can you add in a test for this? There probably doesn't need to be an entirely new test - in `ecommerce/views_test.py` there's an existing `test_checkout_result` that should also check for a valid `PaidCourseRun` record. Looks good otherwise.

rachellougee commented:
@jkachel I've addressed your feedback and added the test. I added some extra commits when tried to resolve merge conflict, but the file changed should reflect the code change in this PR. Please let me know if any additional feedback. Thank you

jkachel reviewed (approved):
LGTM 👍

rhysyngsun commented on ecommerce/api.py:
```diff
@@ -188,6 +189,8 @@ def fulfill_completed_order(
     order.fulfill(payment_data)
     order.save()
 
+    add_fulfilled_order_to_paid_courserun(order)
```
I'd suggest this should be added to `Order.fulfill()` so that we can be sure if that's called from anywhere else that this logic is always executed.

rhysyngsun commented on ecommerce/views/v0/__init__.py:
```diff
@@ -508,6 +508,9 @@ def post(self, request, *args, **kwargs):
             order.save()
 
+            # order is in review state, record it as paid courserun
+            api.add_fulfilled_order_to_paid_courserun(order)
```
Similar to above, I'd suggest this should be added to `Order.review()` .

rhysyngsun commented on ecommerce/api.py:
```diff
@@ -200,6 +203,18 @@ def fulfill_completed_order(
+    """
+    When order is fulfilled or reviewed, record it to paid courserun table if not exist
+    """
+    order = Order.objects.get(id=order.id)
+    if order.STATE.FULFILLED or order.STATE.REVIEW:
+        for run in order.purchased_runs:
+            PaidCourseRun.objects.get_or_create(
+                order=order, course_run=run, user=order.purchaser
+            )
```
This could probably be a method on `Order` so it can just use `self`. I'd also say we already have a property `is_review` so we should use that and also add `is_fulfilled`.

Probably worth adding ones for the other states too and updating any usages, but that's probably best as a separate improvement issue.

rachellougee commented:
@rhysyngsun I've moved this logic to the following places, It's also where we create enrollments, It makes sense to keep it there.

- PendingOrder.fulfill()
- PendingOrder.review()
- ReviewOrder.fulfill()

I had trouble adding the logic to Order model directly due to state issues as it's need to wait until order is saved per discussed.
I ran the pytest after the code changes. Please let me know if any other feedback. Thanks

rhysyngsun commented on ecommerce/models.py:
```diff
@@ -495,6 +510,12 @@ def fulfill(self, payment_data):
         )
 
+        # record all the courserun in the order
+        for run in self.purchased_runs:
```
A few things we could improve overall:

 - Move this code above the `create_run_enrollments` call so that if that API call to openedx fails for some reason we don't get in an inconsistent state.
 - Could you create an abstract base class that subclasses `Order` and that this proxy class would the subclass?
   - Move the `self.transaction.create` and `PaidCourseRun.objects.get_or_create` into their own dedicated functions there
   - Move the `create_run_enrollments()` call into its own dedicated function
   - Move the implementation of `fulfill()` there as well so we don't have duplicate code

rhysyngsun commented on ecommerce/api.py:
```diff
@@ -29,6 +29,7 @@
 
 from mitol.common.utils.datetime import now_in_utc
 from ipware import get_client_ip
+from courses.models import PaidCourseRun
```
Unused import?

rachellougee commented:
@jkachel Do you mind taking another look based on @rhysyngsun 's new feedback? The changes now have new class to reduce code duplication for existing code. Thanks

jkachel reviewed (approved):
Needs a two character change - then good to merge. LGTM 👍
